File: app/src/main/java/com/jann_luellmann/thekenapp/data/view_model/CustomerViewModel.java
package com.jann_luellmann.thekenapp.data.view_model;

import com.jann_luellmann.thekenapp.data.model.Customer;
import com.jann_luellmann.thekenapp.data.model.EventCustomerCrossRef;

import java.util.List;

import androidx.lifecycle.LiveData;

public class CustomerViewModel extends BaseViewModel<Customer> {

    public CustomerViewModel() {
        super();
    }

    public void insert(Customer customer) {
        executor.execute(() -> db.customerDAO().insertAll(customer));
    }

    public void insert(List<Customer> customers) {
        executor.execute(() -> db.customerDAO().insertAll(customers.toArray(new Customer[0])));
    }

    public void update(Customer customer) {
        executor.execute(() -> db.customerDAO().update(customer));
    }

    public LiveData<List<Customer>> findAll() {
        return db.customerDAO().getAll();
    }

    @Override
    public void insert(long eventId, Customer customer) {
        executor.execute(() -> {
            long id = db.customerDAO().insert(customer);
            db.eventCustomerCrossDAO().insertAll(new EventCustomerCrossRef(eventId, id));
        });
    }

    @Override
    public void delete(Customer customer) {
        executor.execute(() -> db.customerDAO().delete(customer));
    }
}

File: app/src/main/java/com/jann_luellmann/thekenapp/data/view_model/CustomerViewModel.kt
package com.jann_luellmann.thekenapp.data.view_model

import androidx.lifecycle.LiveData
import com.jann_luellmann.thekenapp.data.dao.relationship.EventAndCustomerWithDrinksDAO
import com.jann_luellmann.thekenapp.data.model.Customer
import com.jann_luellmann.thekenapp.data.model.EventCustomerCrossRef
import com.jann_luellmann.thekenapp.data.model.EventCustomerDrinkCrossRef

class CustomerViewModel : BaseViewModel<Customer>() {
    fun insert(customer: Customer) {
        executor.execute { db.customerDAO().insertAll(customer) }
    }

    fun insert(customers: List<Customer>) {
        executor.execute { db.customerDAO().insertAll(*customers.toTypedArray()) }
    }

    fun update(customer: Customer) {
        executor.execute { db.customerDAO().update(customer) }
    }

    override fun findAll(): LiveData<List<Customer>> {
        return db.customerDAO().all
    }

    /**
     * Creates a customer for this event and inserts a EventCustomerDrinkCrossRef for each drink
     * in this event.
     */
    override fun insert(eventId: Long, customer: Customer) {
        executor.execute {
            val customerId = db.customerDAO().insert(customer)
            db.eventCustomerCrossDAO().insertAll(EventCustomerCrossRef(eventId, customerId))

            val drinkIds = db.eventDrinkCrossDAO().findAllByIdBlocking(eventId)
            drinkIds.forEach {
                db.eventAndCustomerWithDrinksDAO().insert(EventCustomerDrinkCrossRef(eventId, customerId, it.drinkId))
            }
        }
    }

    override fun delete(customer: Customer) {
        executor.execute { db.customerDAO().delete(customer) }
    }
}